eip-0005.md

```diff
@@ -0,0 +1,170 @@
+`ConstLength`  | `VLQ(UInt)`        | 2                      | Number of items in `Constants` 
+`Constants`    | `Const+`           | (*)                    | Constant serialization bytes (see _Constant Serialization_ section of [ErgoTree Spec](https://ergoplatform.org/docs/ErgoTree.pdf)
+`Template`     | `Expr`             | (**)                   | Contract Template bytes with parameter placeholders (see _Expression Serialization_ section of [ErgoTree Spec](https://ergoplatform.org/docs/ErgoTree.pdf)
+`ParamCount`   | `VLQ(UInt)`        | 1                      | Number of named template parameters
+`Parameters`   | `Parameter+`       | delta: Int             | Typed template parameter (see [Parameter Serialization Format](#parameter-serialization-format) section)
+
+_(*)_
+IR: 
+```
+Seq(
+  Constant(720, SInt), 
+  Constant(SigmaProp(ProveDlog(ECPoint(79be66, 483ada,...))), SSigmaProp)
+)
+```
+Hex: "0x0204a00b08cd0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798"
+
+_(**)_
+IR: 
+```
+SigmaAnd(Seq(
+    BoolToSigmaProp(GE(Height,Plus(SelectField(ExtractCreationInfo(Self),1), ConstantPlaceholder(0,SInt)))),
+    ConstantPlaceholder(1,SSigmaProp)))
+)
+```
+Hex: "ea02d192a39a8cc7a70173007301"
+
+## Parameter Serialization Format
+
+Parameters specify additional meta information for the contract template. The number of
+parameters can be less then equal the number of segregated constants. This is because some
+constant values may not be changed without breaking contract logic and hence they cannot
+be used as template parameters.
+
+Field Name     |  Format            | Example             | Description
+---------------|--------------------|---------------------|-----------
+`NameLength`   | `VLQ(UInt)`        | 5                   | Length of `Name` bytes 
+`Name`         | `Bytes`            | "delta"             | User readable parameter name (string bytes in UTF-8 encoding)
+`DescLength`   | `VLQ(UInt)`        | 14                  | Length of `Description` bytes 
+`Description`  | `Bytes`            | "spending delay"    | User readable parameter description (string bytes in UTF-8 encoding)
+`Type`         | `Type`             | Int                 | ErgoTree type serialized according to section (see _Type Serialization_ section of [ErgoTree Spec](https://ergoplatform.org/docs/ErgoTree.pdf)
+`Placeholder`  | `VLQ(UInt)`        | 0                   | Placeholder index in constants sequence 
+
+Note, the `Placeholder` maps the parameters to the constant placeholders in the
+`Template`.
+
+## JSON Format
+
+In addition to the binary format, the contract templates can be represented using JSON.
+It may serve better for some applications. Below we define a standard JSON representation
+of contract templates.
+
+```json
+{
+  "name": "rewardOutputScript",
+  "description": "holds mining rewards",
+  "constants": "0204a00b08cd0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798",
```

Comment:
How constants are bound to placeholders? Is providing all of them mandatory?

Reply:
Described